File: src/topos/functors/probes/uast/compare.py
"""
UAST Comparison Module
----------------------
Cross-language structural comparison built on top of UAST `kind` values.

The existing `topos.metrics.distance` module is grammar-specific because
it consumes raw tree-sitter node types. This module is the language-agnostic
counterpart: every distance here operates on normalized UAST kinds, so two
implementations of the same algorithm in different languages can be compared
on a single axis.
"""

from __future__ import annotations

from dataclasses import dataclass

from topos.functors.profunctors.distance import DistanceResult, _compute_sequence_distance
from topos.functors.probes.uast.signature import (
    CONTROL_FLOW_KINDS,
    StructuralSummary,
    control_flow_profile,
    structural_summary,
    uast_dfs_kind_sequence,
    uast_kind_histogram,
)


@dataclass(frozen=True)
class UASTComparison:
    """Aggregate cross-language comparison between two UAST roots."""

    kind_distance: float
    edit_distance: DistanceResult
    control_flow_delta: dict[str, int]
    summary_delta: dict[str, int]
    source_summary: StructuralSummary
    target_summary: StructuralSummary

    @property
    def detects_difference(self) -> bool:
        """True if any signature reports a non-zero divergence."""
        if self.kind_distance > 0.0:
            return True
        if self.edit_distance.raw_distance > 0:
            return True
        if any(value != 0 for value in self.control_flow_delta.values()):
            return True
        return any(value != 0 for value in self.summary_delta.values())


def uast_kind_distance(
    source,
    target,
    *,
    include_unknown: bool = True,
) -> float:
    """
    L1 distance between normalized UAST kind histograms.

    Both histograms are normalized to probability distributions so the
    result lies in `[0, 1]` regardless of program size. A return value of
    `0.0` means both programs use the same mix of UAST kinds; `1.0` means
    they share no kinds at all.
    """
    a = uast_kind_histogram(source, include_unknown=include_unknown)
    b = uast_kind_histogram(target, include_unknown=include_unknown)

    total_a = sum(a.values()) or 1
    total_b = sum(b.values()) or 1

    kinds = set(a) | set(b)
    l1 = sum(abs(a.get(k, 0) / total_a - b.get(k, 0) / total_b) for k in kinds)
    # L1 distance between two probability distributions is bounded by 2;
    # halve to get a [0, 1] range matching the rest of the codebase.
    return l1 / 2.0


def uast_edit_distance(
    source,
    target,
    *,
    include_unknown: bool = True,
) -> DistanceResult:
    """
    Tree edit distance over UAST kind sequences (DFS pre-order).

    Reuses the Wagner-Fischer implementation from `topos.metrics.distance`
    so the operation accounting stays consistent with the tree-sitter
    variant.
    """
    source_kinds = uast_dfs_kind_sequence(source, include_unknown=include_unknown)
    target_kinds = uast_dfs_kind_sequence(target, include_unknown=include_unknown)

    distance, ops = _compute_sequence_distance(source_kinds, target_kinds)
    max_size = max(len(source_kinds), len(target_kinds), 1)
    normalized = min(distance / max_size, 1.0)

    return DistanceResult(
        raw_distance=distance,
        normalized_distance=normalized,
        operations=ops,
    )


def _control_flow_delta(source, target) -> dict[str, int]:
    src_profile = control_flow_profile(source)
    tgt_profile = control_flow_profile(target)
    return {kind: tgt_profile[kind] - src_profile[kind] for kind in CONTROL_FLOW_KINDS}


def _summary_delta(
    source_summary: StructuralSummary,
    target_summary: StructuralSummary,
) -> dict[str, int]:
    return {
        "node_count": target_summary.node_count - source_summary.node_count,
        "depth": target_summary.depth - source_summary.depth,
        "declaration_count": (
            target_summary.declaration_count - source_summary.declaration_count
        ),
        "expression_count": (
            target_summary.expression_count - source_summary.expression_count
        ),
        "statement_count": (
            target_summary.statement_count - source_summary.statement_count
        ),
    }


def compare_uast(
    source,
    target,
    *,
    include_unknown: bool = True,
) -> UASTComparison:
    """Run the full UAST comparison suite for a single pair of roots."""
    source_summary = structural_summary(source)
    target_summary = structural_summary(target)
    return UASTComparison(
        kind_distance=uast_kind_distance(
            source, target, include_unknown=include_unknown
        ),
        edit_distance=uast_edit_distance(
            source, target, include_unknown=include_unknown
        ),
        control_flow_delta=_control_flow_delta(source, target),
        summary_delta=_summary_delta(source_summary, target_summary),
        source_summary=source_summary,
        target_summary=target_summary,
    )

File: topos/engine/src/functors/profunctors/uast/compare.rs
//! UAST Comparison — cross-language structural comparison built on top
//! of UAST `kind` values.
//!
//! The AST distance in [`crate::functors::profunctors::ast::compare`]
//! is grammar-specific because it consumes raw tree-sitter node types.
//! This module is the language-agnostic counterpart: every distance
//! here operates on normalized UAST kinds, so two implementations of
//! the same algorithm in different languages can be compared on a
//! single axis.

use std::collections::{HashMap, HashSet};

use crate::functors::probes::uast::signature::{
    control_flow_profile, structural_summary, uast_dfs_kind_sequence, uast_kind_histogram,
    StructuralSummary, CONTROL_FLOW_KINDS,
};
use crate::functors::profunctors::ast::compare::{compute_sequence_distance, DistanceResult};
use crate::graphs::uast::models::UASTNode;

/// Aggregate cross-language comparison between two UAST roots.
#[derive(Debug, Clone, PartialEq)]
pub struct UASTComparison {
    pub kind_distance: f64,
    pub edit_distance: DistanceResult,
    pub control_flow_delta: HashMap<String, i64>,
    pub summary_delta: HashMap<String, i64>,
    pub source_summary: StructuralSummary,
    pub target_summary: StructuralSummary,
}

impl UASTComparison {
    /// True if any structural difference was found.
    pub fn detects_difference(&self) -> bool {
        self.kind_distance > 0.0 || self.edit_distance.raw_distance > 0
    }
}

/// L1 distance between normalized UAST kind histograms.
///
/// Both histograms are normalized to probability distributions so the
/// result lies in `[0, 1]` regardless of program size. A return value
/// of `0.0` means both programs use the same mix of UAST kinds; `1.0`
/// means they share no kinds at all.
pub fn uast_kind_distance(source: &UASTNode, target: &UASTNode, include_unknown: bool) -> f64 {
    let a = uast_kind_histogram(source, include_unknown);
    let b = uast_kind_histogram(target, include_unknown);

    let total_a = a.values().sum::<usize>().max(1) as f64;
    let total_b = b.values().sum::<usize>().max(1) as f64;

    let kinds: HashSet<&String> = a.keys().chain(b.keys()).collect();
    let l1: f64 = kinds
        .iter()
        .map(|k| {
            let av = *a.get(*k).unwrap_or(&0) as f64 / total_a;
            let bv = *b.get(*k).unwrap_or(&0) as f64 / total_b;
            (av - bv).abs()
        })
        .sum();
    // L1 distance between two probability distributions is bounded by 2;
    // halve to get a [0, 1] range matching the rest of the codebase.
    l1 / 2.0
}

/// Tree edit distance over UAST kind sequences (DFS pre-order).
///
/// Reuses the Wagner-Fischer implementation from
/// [`crate::functors::profunctors::ast::compare`] so the operation
/// accounting stays consistent with the tree-sitter variant.
pub fn uast_edit_distance(
    source: &UASTNode,
    target: &UASTNode,
    include_unknown: bool,
) -> DistanceResult {
    let source_kinds = uast_dfs_kind_sequence(source, include_unknown);
    let target_kinds = uast_dfs_kind_sequence(target, include_unknown);

    let (distance, operations) = compute_sequence_distance(&source_kinds, &target_kinds);
    let max_size = source_kinds.len().max(target_kinds.len()).max(1) as f64;
    let normalized = (distance as f64 / max_size).min(1.0);

    DistanceResult {
        raw_distance: distance,
        normalized_distance: normalized,
        operations,
    }
}

fn control_flow_delta(source: &UASTNode, target: &UASTNode) -> HashMap<String, i64> {
    let src_profile = control_flow_profile(source);
    let tgt_profile = control_flow_profile(target);
    CONTROL_FLOW_KINDS
        .iter()
        .map(|&kind| {
            let d = *tgt_profile.get(kind).unwrap_or(&0) as i64
                - *src_profile.get(kind).unwrap_or(&0) as i64;
            (kind.to_string(), d)
        })
        .collect()
}

fn summary_delta(source: &StructuralSummary, target: &StructuralSummary) -> HashMap<String, i64> {
    HashMap::from([
        (
            "node_count".to_string(),
            target.node_count as i64 - source.node_count as i64,
        ),
        (
            "depth".to_string(),
            target.depth as i64 - source.depth as i64,
        ),
        (
            "declaration_count".to_string(),
            target.declaration_count as i64 - source.declaration_count as i64,
        ),
        (
            "expression_count".to_string(),
            target.expression_count as i64 - source.expression_count as i64,
        ),
        (
            "statement_count".to_string(),
            target.statement_count as i64 - source.statement_count as i64,
        ),
    ])
}

/// Run the full UAST comparison suite for a single pair of roots.
pub fn compare_uast(source: &UASTNode, target: &UASTNode, include_unknown: bool) -> UASTComparison {
    let source_summary = structural_summary(source);
    let target_summary = structural_summary(target);
    UASTComparison {
        kind_distance: uast_kind_distance(source, target, include_unknown),
        edit_distance: uast_edit_distance(source, target, include_unknown),
        control_flow_delta: control_flow_delta(source, target),
        summary_delta: summary_delta(&source_summary, &target_summary),
        source_summary,
        target_summary,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graphs::ast::dispatch::parse_source;
    use crate::graphs::uast::models::{NativeRef, SourceSpan};

    fn uast(source: &str, language: &str) -> UASTNode {
        parse_source(source, language, None).unwrap().uast_root
    }

    #[test]
    fn identical_uast_has_zero_distance() {
        let root = uast("def add(a, b):\n    return a + b\n", "python");

        assert_eq!(uast_kind_distance(&root, &root, true), 0.0);
        let edit = uast_edit_distance(&root, &root, true);
        assert_eq!(edit.raw_distance, 0);
        assert_eq!(edit.normalized_distance, 0.0);

        let comparison = compare_uast(&root, &root, true);
        assert!(!comparison.detects_difference());
        assert!(comparison.control_flow_delta.values().all(|&v| v == 0));
        assert!(comparison.summary_delta.values().all(|&v| v == 0));
    }

    #[test]
    fn compare_uast_handles_empty_uast_node() {
        let span = SourceSpan {
            file: None,
            start_byte: 0,
            end_byte: 0,
            start_line: 1,
            start_column: 0,
            end_line: 1,
            end_column: 0,
        };
        let native = NativeRef {
            parser: "test".to_string(),
            parser_version: "0".to_string(),
            node_kind: "module".to_string(),
        };
        let empty = UASTNode {
            kind: "File".to_string(),
            lang: "python".to_string(),
            span,
            native,
            attributes: HashMap::new(),
            children: Vec::new(),
            id: String::new(),
        };
        let other = UASTNode {
            kind: "Unknown".to_string(),
            ..empty.clone()
        };

        let comparison = compare_uast(&empty, &other, true);
        assert_eq!(comparison.kind_distance, 1.0);
        assert_eq!(comparison.edit_distance.raw_distance, 1);
    }
}
